### ISSUE TITLE
EMU tests

### ISSUE THREAD

**Issue by Tom-green1** [ruddock321/y4project#1]
Hi there,

I've had a go at fixing the observables as you said in the meeting however I'm not sure if I have done this right as the HR diagram luminosity and Teff range got slightly worse rather than better. I have included all the plots here. I am mostly not sure about how I have computed the noise as I have used: lum_noise_std = 0.5 * luminosity_07298_ e.g. for luminosity. 

![Image](https://github.com/user-attachments/assets/e1caa7b7-a9b7-42b6-9e10-a6928c2bcffd)

![Image](https://github.com/user-attachments/assets/920bcd27-fa07-4d3f-9b27-6c6bcd6f9acd)

![Image](https://github.com/user-attachments/assets/833b2601-d294-46b3-9e41-362d3655f99c)

![Image](https://github.com/user-attachments/assets/17693c36-8bea-4e68-a280-57eb8f1e4df1)

@grd349 @amaliestokholm

**Comment by grd349** [comment]
Hi Tom, 

Are you still on Campus?

Can you link to the code you are running and I will take a look.  

G

**Comment by Tom-green1** [comment]
Hi Guy,

Sorry I meant to put a link to the code in. Its this bit: 

https://github.com/ruddock321/y4project/blob/Tom/HBM%20stuff/HBM_V8.ipynb

I'm not on campus anymore but I can come back if you are still around.

Thanks

**Comment by grd349** [comment]
Hi Tom, 

I don't understand the following ... 

lum_noise_std = 0.5 * luminosity_07298_
    teff_noise_std = 0.5 * teff_07298_
    dnu_noise_std = 0.5 * dnufit_07298_
    numax_noise_std = 0.5 * numax_07298_
    feh_noise_std = 0.5 * abs(FeH_07298_)


    teff = numpyro.deterministic(
    "teff", 
    teff_07298_ + numpyro.sample("teff_noise", dist.Normal(0, teff_noise_std)))
    lum = numpyro.deterministic(
    "lum", 
    luminosity_07298_ + numpyro.sample("lum_noise", dist.Normal(0, lum_noise_std)))
    dnu = numpyro.deterministic(
    "dnu", 
    dnufit_07298_ + numpyro.sample("dnu_noise", dist.Normal(0, dnu_noise_std)))
    numax = numpyro.deterministic(
    "numax", 
    numax_07298_ + numpyro.sample("numax_noise", dist.Normal(0, numax_noise_std)))
    feh = numpyro.deterministic(
    "feh", 
    FeH_07298_ + numpyro.sample("feh_noise", dist.Normal(0, feh_noise_std)))

What is this supposed to do?

**Comment by grd349** [comment]
I think what you want to do is modify ... 

obs_err = {
    'teff_err': [70],
    'lum_err': [10],
    'dnu_err': [2],
    'feh_err': [0.1],
    'numax_err': [3]
}

to 

obs_err = {
    'teff_err': [70],
    'lum_err': [0.1],
    'dnu_err': [0.1],
    'feh_err': [0.1],
    'numax_err': [0.5]
}

**Comment by Tom-green1** [comment]
In the code before, I was trying to add noise to the observables but completely forgot I had made that. I've just tried using it however and it didn't improve the outcome but I think I know why now. When I am plotting the HR diagram, I have been using the mean values of the model's predictions for massini, fehini, etc but because this track (or any other) doesn't have these exact values the emulator will get it wrong. To clarify, I am using a single value because in the tracks, all the inputs apart from age are just single values repeated. I have had a look at using different indices, e.g. finding the massini value from the model closest to that of the track and then using that for all the others but there's always one or two parameters that will be too far off for the neural network to make sense of it. I was wondering, firstly if this is still a problem with the model, and secondly, if not, how should I select which index to use?

**Comment by grd349** [comment]
>  I was trying to add noise to the observables but completely forgot I had made that

I think it is adding noise in the model - that's not what you want.  You want to add noise to the observable values taken from the track and you want to do this before the inference stage.

> I was wondering, firstly if this is still a problem with the model, and secondly, if not, how should I select which index to use?

I'm sorry, I don't understand.  I don't know which part of the work you are referring to.  

The steps yopu should be doing are something like ... 

1. Pick a point from a track that looks like it could be from M4.
2. Generate the observables using the neural network and add a small amount of noise that is consistent with the observable uncertainty (observed = truth + np.random.randn()*observed_uncertainty).
3. Run the model using the observed values as constraints.

Does that make sense?

**Comment by Tom-green1** [comment]
I think so, should adding the noise look something like this then: obs_err = {
    'teff_err': [70],
    'lum_err': [0.1],
    'dnu_err': [0.1],
    'feh_err': [0.1],
    'numax_err': [0.5]
}

teff_obs = teff_07298_ + np.random.randn() * obs_err['teff_err']
lum_obs = luminosity_07298_ + np.random.randn() * obs_err['lum_err']
dnu_obs = dnufit_07298_ + np.random.randn() * obs_err['dnu_err']
FeH_obs = FeH_07298_ + np.random.randn() * obs_err['lum_err']
numax_obs = numax_07298_ * sun_numax + np.random.randn() * obs_err['numax_err']

obs = {
    'teff': [teff_obs],
    'lum': [lum_obs],
    'dnu': [dnu_obs],
    'feh': [FeH_obs],
    'numax': [numax_obs],
} 

and then using this in the model like: 

if obs is not None:
        numpyro.sample("teff_obs", dist.StudentT(5, teff, obs_err['teff_err'][0]), obs=obs['teff'][0])
        numpyro.sample("lum_obs", dist.StudentT(5, lum, obs_err['lum_err'][0]), obs=obs['lum'][0])
        numpyro.sample("dnu_obs", dist.StudentT(5, dnu, obs_err['dnu_err'][0]), obs=obs['dnu'][0])
        numpyro.sample("feh_obs", dist.StudentT(5, feh, obs_err['feh_err'][0]), obs=obs['feh'][0])
        numpyro.sample("numax_obs", dist.StudentT(5, numax, obs_err['numax_err'][0]), obs=obs['numax'][0])

**Comment by grd349** [comment]
Hi Tom, 

Yes - that looks right.  Throw it into the model and post a link to the updated ipynb in here and I'll look again just to make sure.

**Comment by Tom-green1** [comment]
Hi Guy,

I have re ran it now, the prior predictive is still not working for the observables though, it is still saying NaN or inf axis limits. Heres the link to the code: https://github.com/ruddock321/y4project/blob/Tom/HBM%20stuff/HBM_V8.ipynb

**Comment by grd349** [comment]
Hi Tom, 

```
massini_ = numpyro.deterministic("massini_", 0.8 * numpyro.sample("massini_s", dist.Beta(2, 2)) + 0.7)
    tau_hat = numpyro.deterministic("tau_hat", 4 * numpyro.sample("tau_hat_s", dist.Beta(1.25, 2)) + 1) 

    # Calculate values for joint prior: 
    tau_ms = (2700) * (massini_**-3.15)
    ages_ = numpyro.deterministic("ages_", jnp.minimum(tau_hat * tau_ms, 20000)) 
```

Is that 2700 correct - are the units Myr for age?

**Comment by grd349** [comment]
massini_ = numpyro.deterministic("massini_", 0.8 * numpyro.sample("massini_s", dist.Beta(2, 2)) + 0.7)
    tau_hat = numpyro.deterministic("tau_hat", 0.1 * numpyro.sample("tau_hat_s", dist.Beta(1.25, 2)) + 1) 

    # Calculate values for joint prior: 
    tau_ms = (10,000) * (massini_**-3.15)
    ages_ = numpyro.deterministic("ages_", jnp.minimum(tau_hat * tau_ms,  5,000)) 

Would have been my guess, based on science not results.

**Comment by grd349** [comment]
`x = jnp.hstack(log_transformed_inputs + [fehini_, alphafe_])`

I'm sure you have tested this, but are you sure this works?

**Comment by grd349** [comment]
Otherwise this looks very sensible.  I would try adding in the observables (maybe just Teff and L perhaps), and see if it runs.

**Comment by Tom-green1** [comment]
Hi Guy,

From my prior testing, that value seemed to give the results that best matched the garstec data, with your values it produces a very narrow range of ages with most of them being around 5000. I have tested that the x = jnp.hstack line does work as well. I have got the observables added in and the code runs, its just the prior predictive that it doesnt seem to work for but i think that is due to the neural network as the priors dont seem to be outputting ridiculous values. I have tested with the network as well and even if one of the inputs doesnt match what it expects, the predictions go a bit wild. 

I was going to start experimenting with running multiple stars now unless there are other improvements you can think to add to my model?

Thanks,

Tom

**Comment by grd349** [comment]
Is it working with one star? Can you share those results?

**Comment by Tom-green1** [comment]
I think its working as well as it can with the neural network not being perfect at the moment, the results are here: https://github.com/ruddock321/y4project/blob/Tom/HBM%20stuff/HBM_V8.ipynb
